Apps/Src/SendTritium.c

```diff
@@ -0,0 +1,356 @@
+
+    // LEFT SHIFT
+    cruiseEnableBitmap <<= 1; 
+    regenBitmap <<= 1;
+    cruiseSetBitmap <<= 1;
+}
+
+/**
+ * @brief Converts integer percentage to float percentage
+ * @param percent integer percentage from 0-100
+*/
+extern const float pedalToPercent[];
+static float percentToFloat(uint8_t percent){
+    if(percent > 100){
+        return 1.0f;
+    }
+    return pedalToPercent[percent];
+}
+
+/**
+ * @brief Follows the FSM to update the velocity of the car
+*/
+void Task_SendTritium(void *p_arg){
+    OS_ERR err;
+    Minion_Error_t minion_err;
+    
+    while(1){
+        FSM[state].stateHandler();    // do what the current state does
+        readInputs();   // read inputs from the system
+        FSM[state].stateDecider();    // decide what the next state is
+
+        // Delay of few milliseconds (100)
+        OSTimeDlyHMSM(0, 0, 0, 100, OS_OPT_TIME_HMSM_STRICT, &err);
+        if (err != OS_ERR_NONE){
+            assertOSError(OS_UPDATE_VEL_LOC, err);
+        }
+        // Sets brakelight
+        Minion_Write_Output(BRAKELIGHT, brakelightState, &minion_err);
+    }
+}
+
+// State Handlers & Deciders
+
+/**
+ * Normal Drive State. Accelerator is mapped directly to current setpoint.
+*/
+void NormalDriveHandler(){
+    if(reverseSwitch && !forwardSwitch){
+        velocitySetpoint = -MAX_VELOCITY;
+        currentSetpoint = percentToFloat(accelPedalPercent);
+    }else if(!reverseSwitch && forwardSwitch){
+        velocitySetpoint = MAX_VELOCITY;
+        currentSetpoint = percentToFloat(accelPedalPercent);
+    }
+    else if(!reverseSwitch && !forwardSwitch){
+        currentSetpoint = 0;
+    }
+    brakelightState = false;
+}
+
+void NormalDriveDecider(){
+    if(cruiseSet && cruiseEnable && brakePedalPercent == 0){
+        state = RECORD_VELOCITY;
+    }else if(!cruiseEnable && regenToggle && ChargeEnable){
+        state = ONEPEDAL_DRIVE;
+    }else if(brakePedalPercent > 0){
+        state = BRAKE_STATE;
+    }
+}
+
+/**
+ * Record Velocity State. While pressing the cruise set button, the car will record the observed velocity
+ * into velocitySetpoint.
+*/
+void RecordVelocityHandler(){
+    velocitySetpoint = MotorController_ReadVelocity();
+    brakelightState = true;
+}
+
+void RecordVelocityDecider(){
+    if(cruiseEnable && !cruiseSet){
+        state = POWERED_CRUISE;
+    }else if(!cruiseEnable){
+        state = NORMAL_DRIVE;
+    }else if(brakePedalPercent > 0){
+        state = BRAKE_STATE;
+    }
+}
+
+/**
+ * Powered Cruise State. Continue to travel at the recorded velocity as long as
+ * Observed Velocity <= Velocity Setpoint
+*/
+void PoweredCruiseHandler(){
+    currentSetpoint = 1.0f;
+    brakelightState = false;
+}
+
+void PoweredCruiseDecider(){
+    if(MotorController_ReadVelocity() > velocitySetpoint){
+        state = COASTING_CRUISE;
+    }else if(!cruiseEnable){
+        state = NORMAL_DRIVE;
+    }else if(!cruiseEnable && !cruiseSet){
+        state = RECORD_VELOCITY;
```

Comment:
How would this work exactly? We're in cruise, disable then re enable cruise? It doesn't really hurt anything, but it seems to me like we'd probably go from cruise -> normal -> record velocity whenever the driver turns cruise control off and on again.

Reply:
It was meant to be cruiseEnable && cruiseSet, so that when set is pressed again during a cruise state we go to record velocity.